explainer/effect-generic-trait-declarations.md

```diff
@@ -0,0 +1,503 @@
+associated type should be called `Err` or `Error`.
+
+This might seem minor, but without automation these subtle
+similar-but-not-quite-the-same kinds of differences stand out. The only way to
+ensure that different APIs in different contexts work consistently is via
+automation. And the best automation we have for this is the type system.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Trait definitions
+
+The base of Effect Generic Trait Declarations is the ability to declare traits
+as being generic over effects. This RFC currently only considers the `async`
+effect, but should be applicable to most other effects (modulo `unsafe` and
+`const`, more on that later). The way a trait is defined is by adding a
+`#[maybe(effect)]` notation.  This signals that a trait may be implemented as
+carrying the effect.  For example, a version of the `Into` trait which may or
+may not be `async` would be defined as:
+
+```rust
+#[maybe(async)]
+pub trait Read {
+    #[maybe(async)]
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
+}
+```
+
+## Trait implementations
+
+Traits can be implemented as either async or non-async.  The trait-level
+`#[maybe(async)]` can be thought of as a const-generic bool which determines the
+value of the method-level `#[maybe(async)]` declarations.  So if a trait is
+implemented as `async`, all methods tagged as `#[maybe(async)]` have to be async
+too.
+
+```rust
+/// The base implementation
+impl Read for Reader {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        // ...
+    }
+}
+
+/// The async implementation
+impl async Read for Reader {
+    async fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        // ...
+    }
+}
+```
+
+## Method markers
+
+This RFC only covers trait methods which carry an effect or not. It does not
+cover types which may or may not have effects. The intent is to add this via a
+future extension, so for the scope of this RFC we have to be able to declare
+certain methods as not being generic over effects. This is the default behavior;
+no extra annotations are needed for this.
+
+Taking the `Read` trait example again; the `chain` method returns a type `Chain`
+which implements `Iterator`. Accounting for the `chain` method, the declaration
+of `Read` would be:
+
+```rust
+#[maybe(async)]
+pub trait Read {
+    ...
+
+    // This method is not available for `impl async Read`
+    fn chain<R: Read>(self, next: R) -> Chain<Self, R>
+       where Self: Sized { .. }
+}
+```
+
+Because `chain` is not marked as `maybe(async)`, when implementing `async Read`,
+it will not be available. If a synchronous method has to be available in an
+async context, it should be possible to mark it as `never(async)`, so that it's
+clear it's part of the API contract for the async implementation - and is never
+async.
+
+```rust
+#[maybe(async)]
+pub trait Read {
+    ...
+
+    // This method would be available for `impl async Read`
+    #[not(async)]
```

Comment:
is it `not` or `never`? `never` is used above...

Reply:
It should be `not` - thanks for catching that!